com.unity.ml-agents/Runtime/Agent.cs

```diff
@@ -105,11 +106,26 @@ internal struct AgentAction
         "docs/Learning-Environment-Design-Agents.md")]
     [Serializable]
     [RequireComponent(typeof(BehaviorParameters))]
-    public abstract class Agent : MonoBehaviour
+    public abstract class Agent : MonoBehaviour, ISerializationCallbackReceiver
     {
         IPolicy m_Brain;
         BehaviorParameters m_PolicyFactory;
 
+        /// This code is here to make the upgrade path for users using maxStep
+        /// easier.  We will hook into the Serialization code and make sure that
+        /// agentParameters.maxStep and this.maxStep are in sync.
+        [Serializable]
+        internal struct AgentParameters
+        {
+            public int maxStep;
+        }
+
+        [SerializeField] [HideInInspector]
+        internal AgentParameters agentParameters;
+        [SerializeField] [HideInInspector]
+        internal bool hasUpgradedFromAgentParameters;
+        bool m_MarkDirtFromUpgrade;
```

Comment:
nit: m_MarkDirt*y*FromUpgrade (but not worth doing unless you're going to make other changes)

Reply:
gah! i should change it since it's serialized